AdfDataflowFilePrettifier/Prettifier.cs

```diff
@@ -71,43 +89,115 @@ private static (string, string, string) SplitFileIntoInterestingSections(string
+                var blockContent = ReadStringBetween(fileContents, block.BlockContentStartIndex, block.BlockContentEndIndex);
 
-            uglierFile = uglierFile.Replace(OuterMagicIndicatorCharacter + "\n", "");
-            uglierFile = uglierFile.Replace(CRMagicIndicatorCharacter + "\n", "\\r\\n");
-            uglierFile = uglierFile.Replace(CRMagicIndicatorCharacter + "\r", "\\r");
-            uglierFile = uglierFile.Replace(InnerMagicIndicatorCharacter + "\n", "\\n");
-            uglierFile = uglierFile.Replace(InnerMagicIndicatorCharacter + "\t", "\\t");
+                uglyFileBuilder.Append(newContentUpToBlockStart);
+                uglyFileBuilder.Append(UglifyString(blockContent));
 
-            return uglierFile;
+                endOfLastBlock = block.EndOfBlockEndMarkerIndex;
+            }
+
+            // No more blocks in the file, read rest of file into string
+            uglyFileBuilder.Append(fileContents.Substring(endOfLastBlock));
+
+            return uglyFileBuilder.ToString();
+        }
+
+        private static IEnumerable<BlockIndices> GetAllBlockIndices(string str)
+        {
+            var nextBlockIndicies = GetNextBlockIndices(str, 0);
+            while (nextBlockIndicies != null)
+            {
+                yield return nextBlockIndicies;
+                nextBlockIndicies = GetNextBlockIndices(str, nextBlockIndicies.EndOfBlockEndMarkerIndex);
+            }
         }
 
+        private static BlockIndices GetNextBlockIndices(string str, int previousBlockEndIndex)
+        {
+            var startOfBlockStartMarkerIndex = str.IndexOf(BlockStartString, previousBlockEndIndex);
+
+            if (startOfBlockStartMarkerIndex == -1)
+            {
+                return null;
+            }
+
+            var blockContentStartIndex = startOfBlockStartMarkerIndex + BlockStartString.Length;
+            var blockContentEndIndex = str.IndexOf(BlockEndString, blockContentStartIndex);
+
+            if (blockContentEndIndex == -1)
+            {
+                throw new Exception("Error uglifying: Cannot find matching end marker for start marker");
+            }
+
+            var endOfBlockEndMarkerIndex = blockContentEndIndex + BlockEndString.Length;
+
+            return new BlockIndices
+            {
+                StartOfBlockStartMarkerIndex = startOfBlockStartMarkerIndex,
+                BlockContentStartIndex = blockContentStartIndex,
+                BlockContentEndIndex = blockContentEndIndex,
+                EndOfBlockEndMarkerIndex = endOfBlockEndMarkerIndex
+            };
+        }
+
+        private static string ReadStringBetween(string str, int startIndex, int endIndex)
+        {
+            return str.Substring(startIndex, endIndex - startIndex);
+        }
+
+        private static string PrettifyString(string uglyString)
+        {
+            var prettyierString = uglyString;
+
+            prettyierString = prettyierString.Replace("\\r", "[[CRLF]]");

```

Comment:
This is my suggestion for replacing `\\r`. It avoids use of control characters to make things more explicit for someone doing the merge / diff.

`\r\n` appears very rarely, so this shouldn't affect the general readability of the script, but it also makes things a lot less magic, and gives a person doing manual merges the ability to add in `\r` if they wish to.

There's a tiny risk that a `[[CRLF]]` might already be present in the file, or that someone might accidentally edit it to eg `[[CRLF]` or something and it to remain in the file. If we'd want to be extra careful and willing to add in a tiny bit more magic, we could add a control character in the middle of the string to reduce the likelihood of false positives.

Also, I chose [[CRLF]] instead of [[CR]] because the former is more clearly an end of line marker and more intuitive (even if it's not strictly correct).

Interested to hear your thoughts.

Reply:
Yes, I think that's definitely the right solution.

Thanks for the explanation on CR/CRLF - that was the first thought I had.
Are you concerned that people who would otherwise recognise `[[CRLF]]` might be interpret `[[CR]]` as something else (due to lack of prompt to consider it a line ending?)

If so that seems reasonable.

Reply:
Other than that, I'm happy with this. Thank you!

Reply:
> Are you concerned that people who would otherwise recognise [[CRLF]] might be interpret [[CR]] as something else (due to lack of prompt to consider it a line ending?)

Yes, exactly. If you weren't aware of what was going on, `[[CR]]` doesn't immediately say to me "line ending weirdness", and could conceivably be something in the ADF scripting world.

Would, say, `[[CarriageReturn]]` be preferred to `[[CRLF]]`? I think I have a slight preference for `[[CRLF]]`, but up to you?

Let me know. Otherwise, I think this is ready to merge if/when you're happy. Thanks for all your help with reviewing this! 😀

Reply:
We could also just leave it as `\\r` of course, although I think `[[CRLF]]` is a slight preference. Still, very happy to give you a deciding vote :D

Reply:
ooh ... leaving it as `\\r` is kinda tempting.
But I think I prefer `[[CRLF]]` since it's clearer that it's a marker we've put in deliberately (and not, say, a bug in the prettifier).

So I think we're good to leave it as it currently is.